### ISSUE TITLE
0xShax2nk.in - Borrowers can avoid late fees by combining existed laons using applyCombine function

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2024-03-zivoe-judging#364]
0xShax2nk.in

medium

# Borrowers can avoid late fees by combining existed laons using applyCombine function

## Summary

The `applyCombine()` function in the OCC_Modular contract is designed to combine multiple active loans into a single new loan. The function aggregates the `principalOwed` from each loan specified in the combinations mapping. However, it does not account for any late fees that may have accrued on these individual loans if a payment has been missed but the grace period has not yet expired.

## Vulnerability Detail

The issue arises with the handling of late fees. In the current implementation, if a borrower has missed a payment on any of their loans, the `applyCombine()` function does not include the late fees that have started to accrue due to the missed payment. The function only considers the `principalOwed` of each loan and overlooks the additional late fees that should be accumulating once the `paymentDueBy` date has passed but before the `gracePeriod` has ended.

This creates a issue where a borrower, who has an active loan with a missed payment within the grace period, can initiate a loan combination through `applyCombine()`. By doing so, they can effectively reset the payment schedule of the delinquent loan without having to pay the late fees that they have incurred. The new combined loan will only reflect the sum of the principals of the individual loans, and the late fees that were accruing on the missed payment are not transferred to the new loan.

## Impact

The current implementation of the `applyCombine` function could allow borrowers to avoid paying late fees by combining loans that have missed payments but are still within their grace period. Since late fees are not included in the new combined loan's principal, this can lead to a loss of expected revenue for the lender and can be exploited by borrowers to escape financial penalties for late payments.

## Code Snippet

https://github.com/sherlock-audit/2024-03-zivoe/blob/main/zivoe-core-foundry/src/lockers/OCC/OCC_Modular.sol#L742

## PoC

Consider the following scenario where a borrower has three active loans with the following states:

 -   Loan 1: On-time payments.
 -  Loan 2: Payment missed, but within grace period.
 -  Loan 3: On-time payments.

The borrower realizes that they have missed a payment for Loan 2 and late fees are accruing. They request an off-chain approval from the underwriter to combine their loans. The underwriter approves the combination by calling `approveCombine`. The borrower then calls `applyCombine` before the grace period for Loan 2 ends.

According to the current contract logic, the new combined loan will only include the `principalOwed` of Loan 1, Loan 2, and Loan 3, without the late fees that should have been accruing on Loan 2. As a result, the borrower effectively avoids paying the late fee on Loan 2.

## Tool used

Manual Review

## Recommendation

To resolve this issue, the contract should be updated to include the calculation and inclusion of any accrued late fees when combining loans. This can be done by calling the `amountOwed` function for each loan within the `applyCombine` function and including the late fee in the new combined loan's principal. The contract logic should be as follows:

1. Within the `applyCombine` function, for each loan ID in the combination:

-    Call `amountOwed` to get the current total owed, including any late fees.
-    Add the total owed to the combined principal amount.

2. Update the state of the individual loans to Combined only after ensuring that all financial obligations (including late fees) have been accounted for in the new combined loan.

3. Create the new combined loan with the updated principal that includes late fees.

By implementing these changes, the contract will ensure that late fees are properly handled during the loan combination process, preventing borrowers from avoiding late fees and maintaining the integrity of the lending system.

Duplicate of #95
